### PR TITLE
Add a test for a bug in the switcher, alongside a fix

### PR DESCRIPTION
**Pull request by saaramar**
The switcher does not check the validity of the target entry point. Add a check for that and implement a test.

### CODE REVIEW

File: sdk/core/switcher/entry.S [hunk 1]
```diff
@@ -311,6 +315,9 @@ exception_entry_asm:
 	reloadRegisters c1, cgp, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14, c15, csp
 	mret
 
+.Lforce_unwind_with_return_values:
```
[the remark this answers is not in the document]
Comment (saaramar) [reply sdk/core/switcher/entry.S]: Thanks! Done in this commit https://github.com/microsoft/cheriot-rtos/pull/8/commits/6d7ea17057dc4677ebd77c188f069514acd3c008

File: sdk/core/switcher/entry.S [hunk 2]
```diff
@@ -311,6 +315,16 @@ exception_entry_asm:
 	reloadRegisters c1, cgp, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14, c15, csp
 	mret
 
+.Linvalid_entry:
+// Mark this threads as in the middle of a forced unwind.
```
Comment (davidchisnall) [inline sdk/core/switcher/entry.S:319]: Can you expand this to explain *why* (we want to resume rather than unwind if the compartment doesn't have an error handler)?

File: tests/compartment_calls-test.cc [hunk 3]
```diff
@@ -40,3 +40,9 @@ void test_compartment_call()
 	  compartment_call_inner(value, value, &value, value, &value, value, value);
 	TEST(ret == 0, "compartment_call_inner returend {}", ret);
 }
+
+void test_compartment_call()
+{
+	test_number_of_arguments();
+	test_incorrect_export_table(NULL);
```
Comment (davidchisnall) [inline tests/compartment_calls-test.cc:47]:
```suggestion
	test_incorrect_export_table(nullptr);
```

File: tests/compartment_calls_inner.cc [hunk 4]
```diff
@@ -10,6 +10,12 @@
 
 using namespace CHERI;
 
+extern "C" ErrorRecoveryBehaviour
+compartment_error_handler(ErrorState *frame, size_t mcause, size_t mtval)
+{
+	return ErrorRecoveryBehaviour::ForceUnwind;
+}
+
```
Comment (davidchisnall) [inline tests/compartment_calls_inner.cc:18]: This should be the default, do we need it to exist?

Comment (davidchisnall) [reply tests/compartment_calls_inner.cc:18]: I see, we don't unwind here because it's treated as a failure in the callee without the handler.  Please add a comment explaining this.

Comment (saaramar) [reply tests/compartment_calls_inner.cc:18]: Exactly. Great idea, I'm adding a comment and another inner compartment to tests both cases.

Comment (davidchisnall) [reply tests/compartment_calls_inner.cc:18]: It would also be good to update the error handling doc for some of these corner cases.  For example, how are each of the following handled:

 - Stack exhaustion
 - Trusted stack exhaustion
 - Invalid return addresses
 - Invalid callee addresses

We don't yet correctly handle the stack exhaustion case, but I'm finishing up the compiler change to generate the correct metadata.

Comment (saaramar) [reply tests/compartment_calls_inner.cc:18]: True, these corner cases are important and we have a good picture of how they should look like. I'll do that in a separate PR.

File: tests/compartment_calls_inner.cc [hunk 5]
```diff
@@ -95,4 +101,17 @@ int compartment_call_inner(int        x0,
...
+{
+	/*
+	 * Trigger a cross-compartment call with an invalid export entry.
+	 */
+
+	debug_log("test an incorrect export table entry");
+
+	fn();
+
+	TEST(false, "Should be unreachable");
```
[the remark this answers is not in the document]
Comment (saaramar) [reply tests/compartment_calls_inner.cc:116]: Yep, already done; going to commit this in a few minutes :)

File: sdk/core/switcher/entry.S [hunk 6]
```diff
@@ -336,10 +349,10 @@ exception_entry_asm:
 	LoadCapPCC     cs0, compartment_switcher_sealing_key
 	// ca2 at this point was loaded by .Lpop_trusted_stack_frame from the pcc
 	// in the trusted stack and so should always be sealed as a sentry type.
-	cgettype       gp, cra
+	cgettype       gp, ca2
```
Comment (davidchisnall) [inline sdk/core/switcher/entry.S]: This looks like it's accidentally reverting a fix?

**Review by davidchisnall — APPROVED** [review]
Looks good, thanks!
